win32: win+up/down hotkeys improvements

pavelxdd opened the pull request:
Continuation of #3894

A few improvements to make borderless/fullscreen window maximize/minimize behaviour to be more consistent with the normal windows:

- <kbd>Win</kbd>+<kbd>↑</kbd> will turn the fullscreen on for a borderless window
- <kbd>Win</kbd>+<kbd>↓</kbd> will minimize the fullscreen window to taskbar 
(unless it's maximized - in that case it will restore back to maximized window - for example Firefox behaves the same way)
- <kbd>Win</kbd>+<kbd>↓</kbd> will minimize the normal window regardless whether it has borders or not (that's already implemented in the previous commit)

I introduced the new variable `toggle_fs` to avoid modifying the `opts->fullscreen` directly (which wouldn't be thread-safe I assume?)

pavelxdd commented:
Found and fixed a few issues while I was testing it:

- An issue from the previous PR: if switch a maximized window to fullscreen and press <kbd>Win</kbd>+<kbd>↓</kbd>, the fullscreen state won't switch to `off`. Fixed by processing SC_RESTORE event.

- Fixed inconsistencies with handling maximize/minimize events for a maximized borderless window.

pavelxdd commented on video/out/w32_common.c:
```diff
@@ -1543,6 +1582,10 @@ static int gui_thread_control(struct vo_w32_state *w32, int request, void *arg)
             reinit_window_state(w32);
         return VO_TRUE;
+    case VOCTRL_GET_FULLSCREEN:
+        *(int *)arg = w32->toggle_fs ^ w32->current_fs;
```
It's the same as `*(int *)arg = w32->toggle_fs ? !w32->current_fs : w32->current_fs`
I can rewrite it if you want.

wm4 replied:
You could simplify it as `w32->toggle_fs != w32->current_fs` (which is exactly the same because both fields are C99 bools).

I mean logically. toggle_fs is "whether the current fullscreen state needs to be switched", and current_fs is the real state. I guess my question is mostly about how toggle_fs is supposed to work.

pavelxdd replied:
You're right, thanks.

wm4 commented on video/out/w32_common.c:
```diff
@@ -1543,6 +1582,10 @@ static int gui_thread_control(struct vo_w32_state *w32, int request, void *arg)
         return VO_TRUE;
+    case VOCTRL_GET_FULLSCREEN:
+        *(int *)arg = w32->toggle_fs ^ w32->current_fs;
+        reinit_window_state(w32);
```
So this can't be right here. It shouldn't change any state in a VOCTRL_GET_FULLSCREEN request. The intention of the request is to retrieve the state. Though due to not yet cleaned up issues, the vo_control code down the call stack might overwrite any values in opts-> and update to the current values the frontend (i.e. playloop, command.c etc.) knows. The purpose of the GET request is to update the values in the frontend, so hitting the `F` key and similar will toggle the fullscreen state correctly.

So I'm not sure if you've considered this when writing the code. This stuff is pretty confusing and not very well-defined after all.

pavelxdd replied:
Actually I considered it and asked you where I should call `reinit_window_state` in the previous PR.
But I couldn't find another place where it could work like intended.

pavelxdd replied:
The only way I see is to move `reinit_window_state` above the message queue handler and call right after the `signal_events`. Does that sound okay?

wm4 replied:
Depends what the intention is. VOCTRL_GET_FULLSCREEN is only for copying the real fullscreen state to the frontend, if it has changed on its own in the win32 thread. To which line would you move the call?

pavelxdd replied:
The intention is to call `reinit_window_state` to update the fullscreen state and all other related variables.
Move it right above the `WndProc`.

What if I signal the `VO_EVENT_WIN_STATE` event instead, will it call the `reinit_window_state`?

wm4 replied:
> What if I signal the VO_EVENT_WIN_STATE event instead, will it call the reinit_window_state?

No, how would it even do that?

pavelxdd replied:
Yeah, I see. This looks overly complicated already. Maybe I should forget about this fullscreen toggle stuff, and just fix this issue with the previous PR and that's all?
As for the Win+Up hotkey, it could just maximize the window, instead of switching to fullscreen.

wm4 replied:
I think maybe it would be better if e.g. VOCTRL_FULLSCREEN copied the option value to toggle_fs, and having `reinit_window_state()` ignore the option value. I think that could possibly make the code a bit easier to follow.

pavelxdd replied:
As far as I can see `VOCTRL_FULLSCREEN` is only triggered on hotkey.

wm4 replied:
> As far as I can see VOCTRL_FULLSCREEN is only triggered on hotkey/command

Yes, that's as designed. It'll also be triggered if the user changes the fullscreen option in any other way. The only exception is if the fullscreen flag is to be updated from the VO settings, e.g.:
- user changes fullscreen "from outside" (like with win+up I guess?)
- win32 code notices this, updates its internal fullscreen state, sets VO_EVENT_FULLSCREEN_STATE
- in reaction to VO_EVENT_FULLSCREEN_STATE playloop.c sends VOCTRL_GET_FULLSCREEN
- and writes the receives value to the fullscreen option variable (without sending VOCTRL_FULLSCREEN)

pavelxdd replied:
I think I got an idea how to do it, thanks!

pavelxdd commented:
So I couldn't do it after all, reverted all changes and only fixed an issue that
appeared after #3894 : 
> if switch a maximized window to fullscreen and press Win+Down, 
the fullscreen state won't switch to off. Fixed by processing SC_RESTORE event.

Sorry for wasting your time, @wm4. Win+Up just maximizes the window, it's not a
real fullscreen, so I can't just set the `current_fs` variable to true and pretend
it's a fullscreen mode.

Maybe I'll return to this later if there ever would be a more easy way to change
vo properties.

wm4 commented:
Depends whether `F` is supposed to be able to switch this "not real" fullscreen.

pavelxdd commented:
Well, I found another way to call `reinit_window_state` on toggling fullscreen. Does it look good now?

pavelxdd commented:
For some reason the events 'SC_MAXIMIZE', 'SC_MINIMIZE' and 'SC_RESTORE' doesn't work on Windows 7 for me, so the Win+Down/Up hotkeys don't work as intended for a borderless window.

pavelxdd commented:
Windows 7 and Windows 8 don't send `WM_SYSCOMMAND` on minimizing/maximizing via Win+Down/Up, so I don't know how to handle these events *before* they happen.

Reverted these changes again and provided a fix for updating the fullscreen on restoring the maximized window from fullscreen.

Win+Up/Down along with Win+Left,Right support can be added later as a part of implementing Aero Snap support for a borderless window.

wm4 commented:
Whatever happened to this? CC @rossy

rossy commented:
@wm4 @pavelxdd Sorry, I forgot about this.

> Reverted these changes again and provided a fix for updating the fullscreen state on restoring the maximized window from fullscreen.

It seems to fix this issue for me. Without this PR, if I enter fullscreen and press Win+Down, the window gets stuck without a border. With this PR, pressing Win+Down restores the window to the non-fullscreen state. Assuming the ``toggle_fs`` stuff is fine (ping @wm4) I guess this is okay.

wm4 commented:
Honestly not sure, but maybe ok. I think the fullscreen state should be managed differently anyway (but this would require changes in the player frontend and other backends as well). So fine with me.

rossy commented:
I think I'll merge this then. If fullscreen handling changes in the frontend, this can be changed too.

rossy commented:
Merged as 16c5d43322
